File: Data Structures & Algorithms/island-perimeter/submission-0.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]

        res = 0
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 1:
                    for dr, dc in directions:
                        if r + dr < 0 or r + dr >= ROWS or c + dc < 0 or c + dc >= COLS or grid[r + dr][c + dc] == 0:
                            res += 1
        return res

File: Data Structures & Algorithms/number-of-enclaves/submission-0.cpp
class Solution {
private:
    int rows, cols;
    vector<pair<int, int>> directions;
    vector<vector<bool>> visited;

    bool borders(int r, int c) {
        return r == 0 || c == 0 || r == rows - 1 || c == cols - 1;
    }

    int dfs(vector<vector<int>>& grid, int r, int c) {
        if (r < 0 || c < 0 || r >= rows || c >= cols || grid[r][c] == 0 || visited[r][c]) return 0;

        visited[r][c] = true;
        int res = 1;
        for (auto& [dr, dc] : directions) {
            res += dfs(grid, r + dr, c + dc);
        }
        return res;
    }
public:
    int numEnclaves(vector<vector<int>>& grid) {
        rows = grid.size(); cols = grid[0].size();
        directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        visited = vector<vector<bool>>(rows, vector<bool>(cols, false));

        int land = 0, borderLand = 0;
        for (int r = 0; r < rows; ++r) {
            for (int c = 0; c < cols; ++c) {
                land += grid[r][c];
                if (grid[r][c] == 1 && !visited[r][c] && borders(r, c)) {
                    borderLand += dfs(grid, r, c);
                }
            }
        }

        return land - borderLand;
    }
};